core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java

```diff
@@ -73,182 +72,219 @@ public class DefaultBlockWorkerClient implements BlockWorkerClient {
    * @param alluxioConf Alluxio configuration
    */
-  public DefaultBlockWorkerClient(UserState userState, GrpcServerAddress address,
-      AlluxioConfiguration alluxioConf) throws IOException {
-    RetryPolicy retryPolicy = RetryUtils.defaultClientRetry();
-    UnauthenticatedException lastException = null;
-    // TODO(feng): unify worker client with AbstractClient
-    while (retryPolicy.attempt()) {
-      try {
-        // Disables channel pooling for data streaming to achieve better throughput.
-        // Channel is still reused due to client pooling.
-        mStreamingChannel = GrpcChannelBuilder.newBuilder(address, alluxioConf)
-            .setSubject(userState.getSubject())
-            .setNetworkGroup(GrpcNetworkGroup.STREAMING)
-            .build();
-        mStreamingChannel.intercept(new StreamSerializationClientInterceptor());
-        // Uses default pooling strategy for RPC calls for better scalability.
-        mRpcChannel = GrpcChannelBuilder.newBuilder(address, alluxioConf)
-            .setSubject(userState.getSubject())
-            .setNetworkGroup(GrpcNetworkGroup.RPC)
-            .build();
-        lastException = null;
-        break;
-      } catch (StatusRuntimeException e) {
-        close();
-        throw AlluxioStatusException.fromStatusRuntimeException(e);
-      } catch (UnauthenticatedException e) {
-        close();
-        userState.relogin();
-        lastException = e;
-      }
-    }
-    if (lastException != null) {
-      throw lastException;
-    }
-    mStreamingAsyncStub = BlockWorkerGrpc.newStub(mStreamingChannel);
-    mRpcBlockingStub = BlockWorkerGrpc.newBlockingStub(mRpcChannel);
-    mRpcAsyncStub = BlockWorkerGrpc.newStub(mRpcChannel);
+  public DefaultBlockWorkerClient(ClientContext context, GrpcServerAddress address,
+                                  AlluxioConfiguration alluxioConf) {
+    // BlockWorkerClient is typically used inside a BlockInStream
+    // to fetch block chunk. BlockInStream has its own retry mechanism, so we don't
+    // retry at this level so as not to interfere with the retry of our higher-level
+    // callers.
+    // An instance where retries interfere with each other is when we and our calling
+    // BlockInStream both adopted a timer-based retry, then all the time could be wasted
+    // in our inner retry, and the outer stream has no chance of retrying a different client.
+    super(context, null, () -> new CountingRetry(0));
+    // the server address this client connects to. It might not be an IP address
     mAddress = address;
     mRpcTimeoutMs = alluxioConf.getMs(PropertyKey.USER_RPC_RETRY_MAX_DURATION);
     mTracker = DETECTOR.track(this);
   }
 
+  @Override
+  protected void beforeConnect() {}
+
+  @Override
+  protected void afterConnect() {
+    // these are alias to the channel built in parent class
+    mStreamingChannel = Preconditions.checkNotNull(
+            mAdditionalChannels.get(GrpcNetworkGroup.STREAMING));
```

Comment:
The channels are constructed in the super class, but re-assigned/aliased to a field in a child class - this is very a confusing and convoluted solution.

Reply:
We could get rid of the alias pretty easily. They're here just to be compatible with the original code.